### DISCUSSION TITLE
Can pprint/pretty be temporarily be disabled?

### DISCUSSION THREAD

**Discussion by aroberge** [Textualize/rich#1603] [Q&A] [answered]
This question came as the result of an issue [1] filed in one of my projects.

When using IPython (on its own), it is possible to toggle on and off the pretty printing.  For example, doing

```
dir(__builtins__)
In [1]: dir(__builtins__)
Out[1]:
['ArithmeticError',
 'AssertionError',
 'AttributeError',
 'BaseException',
 'BlockingIOError',
 'BrokenPipeError',
 'BufferError',
 'BytesWarning',
 'ChildProcessError',
 'ConnectionAbortedError',
 'ConnectionError',
...
```
outputs names one at a time on a line.  However, after turning off the pretty printing with the magic command `%pprint`,  the display is shown without pretty formatting:
```
In [2]: %pprint
Pretty printing has been turned OFF

In [3]: dir(__builtins__)
Out[3]: ['ArithmeticError', 'AssertionError', 'AttributeError', 'BaseException', 'BlockingIOError', 'BrokenPipeError', 'BufferError', 'BytesWarning', 'ChildProcessError', 'ConnectionAbortedError', 'ConnectionError', 'ConnectionRefusedError', 'ConnectionResetError', 'DeprecationWarning', 'EOFError', 'Ellipsis', 'EnvironmentError', 'Exception', 'False', 'FileExistsError', ...
```

When using Rich with IPython, attempting to use the magic toggle `%pprint` results in a traceback.
In [4]: from rich import pretty

In [5]: pretty.install()

In [6]: %pprint
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-1efe34cfb034> in <module>
----> 1 get_ipython().run_line_magic('pprint', '')
    306         """Toggle pretty printing on/off."""
    307         ptformatter = self.shell.display_formatter.formatters['text/plain']
--> 308         ptformatter.pprint = bool(1 - ptformatter.pprint)
    309         print('Pretty printing has been turned',
    310               ['OFF','ON'][ptformatter.pprint])

AttributeError: 'BaseFormatter' object has no attribute 'pprint'
...

Looking at the source code for Rich, this is not surprising since I can see that it monkeypatches IPython's code, substituting its own pretty printer.

I am wondering if there is a way to (temporarily) disable Rich's pretty printer once it is installed.

[1] https://github.com/friendly-traceback/friendly/issues/14

**Comment by willmcgugan** [comment, marked as answer]
I'm sure it is possible. I'm assuming it is toggling `pprint` which is a boolean. I'll look in to it.

**Reply by aroberge** [reply]
The problem is that Rich's pretty replacement for IPython's formatter does not have such a boolean.  I have found a workaround solution for my project. I will try to add a link to the code I use in the next reply.

**Reply by aroberge** [reply]
https://github.com/friendly-traceback/friendly/blob/637904b84b3ff9043d3a605d8cf8a6f580474d66/friendly/ipython.py#L54 is what I use to disable Rich's pretty which currently, in my program, has to be reenabled later as a separate step (and not simply by toggling the pprint boolean).  It would be nice if Rich's pretty would have a similar toggle (both for ipython and in other environments).

**Reply by willmcgugan** [reply]
Thanks, I think I can do something similar within Rich.

**Reply by willmcgugan** [reply]
This line is a bit puzzling (I'm assuming its pulled from IPython):

```python
formatter.pprint = bool(1 - formatter.pprint)
```

I wonder why not this:

```python
formatter.pprint = not formatter.pprint
```

**Reply by aroberge** [reply]
Yes, it was straight from IPython. I remember thinking I should replace it but, after I got things working, I moved on to other things forgetting to come back to this.
